File: Top 100 Codes/04-Important Codes related to Arrays/23-Determine can all numbers of an array be made equal/Determinecanallnumbersofanarraybemadeequal.cpp
#include <iostream>
#include <algorithm> // For sort()
#include <numeric>   // Use std::__gcd for C++14 or earlier

using namespace std;

bool canMakeEqual(int arr[], int n) {
    if (n == 1) 
        return true; // A single number is always equal to itself

    // Sort the array to get the smallest element
    sort(arr, arr + n);

    // Calculate GCD of all differences
    int gcdValue = arr[1] - arr[0];
    for (int i = 2; i < n; i++) {
        gcdValue = __gcd(gcdValue, arr[i] - arr[0]); // Use __gcd()
    }

    return gcdValue != 0;  // If GCD is nonzero, they can be made equal
}

int main() {
    int arr[] = {6, 10, 14}; 
    int n = sizeof(arr) / sizeof(arr[0]);

    if (canMakeEqual(arr, n)) {
        cout << "Yes, all numbers can be made equal.\n";
    } else {
        cout << "No, all numbers cannot be made equal.\n";
    }

    return 0;
}

File: Top 100 Codes/04-Important Codes related to Arrays/23-Determine can all numbers of an array be made equal/Determinecanallnumbersofanarraybemadeequal.py
from math import gcd
from functools import reduce

def can_make_equal(arr):
    if len(arr) == 1:
        return True  # A single number is always equal to itself

    arr.sort()  # Sort the array to find the smallest element
    differences = [arr[i] - arr[0] for i in range(1, len(arr))]

    # Compute the GCD of all differences
    gcd_value = reduce(gcd, differences)

    return gcd_value != 0  # If GCD is nonzero, numbers can be made equal

# Example test case
arr = [6, 10, 14]  # Example test case
if can_make_equal(arr):
    print("Yes, all numbers can be made equal.")
else:
    print("No, all numbers cannot be made equal.")
